### PR TITLE
Website: 0.12 config language docs edits

### PR DESCRIPTION
**Pull request by nfagerlund**
Sorry: this is large. I talked through most of the more radical edits with Martin ahead of time, and hopefully I remembered to actually go with the solutions we jointly decided on! 

There are a few sections where I could use some extra attention to the accuracy of my edits; I'll call those out in comments.

### CODE REVIEW

File: website/docs/configuration/expressions.html.md [hunk 1]
```diff
@@ -568,29 +570,40 @@ resource "aws_security_group" "example" {
...
+- The label of the dynamic block (`"ingress"` in the example above) specifies
+  what kind of nested block to generate.
+- The `for_each` argument provides the complex value to iterate over.
+- The `iterator` argument (optional) sets the name of a temporary variable
+  that represents the current element of the complex value. If omitted, the name
+  of the variable defaults to the label of the `dynamic` block (`"ingress"` in
+  the example above).
+- The `labels` argument (optional) is a list of strings that specifies the block
+  labels, in order, to use for each generated block. You can use the temporary
+  iterator variable in this value.
```
Comment (nfagerlund) [inline website/docs/configuration/expressions.html.md]: (checking to make sure I got this extra argument correct)

File: website/docs/configuration/index.html.md [hunk 2]
```diff
@@ -16,23 +16,52 @@ describing an intended goal rather than the steps to reach that goal.
...
 which creates a larger unit of configuration. A resource describes a single
 infrastructure object, while a module might describe a set of objects and the
 necessary relationships between them in order to create a higher-level system.
 
 A _Terraform configuration_ consists of a _root module_, where evaluation
-begins, along with a tree of child modules created when one module references
+begins, along with a tree of child modules created when one module calls
 another.
 
+## Arguments, Blocks, and Expressions
```
Comment (nfagerlund) [inline website/docs/configuration/index.html.md]: This is the bit where I was talking about adding a capsule version of the syntax outline to index.html, so that readers can stay away from syntax.html if they want.

File: website/docs/configuration/providers.html.md [hunk 3]
```diff
@@ -172,14 +178,26 @@ configuration. For providers that have no required configuration arguments, the
 implied _empty_ configuration is considered to be the _default_ provider
 configuration.
 
-Resources are normally associated with the default provider configuration
-inferred from the resource type name. For example, a resource of type
-`aws_instance` uses the _default_ (un-aliased) `aws` provider configuration
-unless otherwise stated.
+### Referring to Alternate Providers
```
Comment (nfagerlund) [inline website/docs/configuration/providers.html.md]: (this, below, was a bit that I remember having trouble with; could use confirmation that I got it right.)

File: website/docs/configuration/style.html.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+layout: "docs"
+page_title: "Style Conventions - Configuration Language"
+sidebar_current: "docs-config-style"
+description: |-
+  The Terraform language has some idiomatic style conventions, which we
+  recommend users always follow for consistency between files and modules
+  written by different teams.
+---
+
+# Style Conventions
```
Comment (nfagerlund) [inline website/docs/configuration/style.html.md]: (broke this into a separate page for convenience and legibility.)

File: website/docs/configuration/syntax.html.md [hunk 5]
```diff
@@ -30,23 +30,39 @@ details. If you are interested, you can find a full definition of HCL
...
+The identifier before the equals sign is the _argument name_, and the expression
+after the equals sign is the argument's value.
+
+The context where the argument appears determines what value types are valid
+(for example, each resource type has a schema that defines the types of its
+arguments), but many arguments accept arbitrary
+[expressions](./expressions.html), which allow the value to
+either be specified literally or generated from other values programmatically.
+
+-> **Note:** Terraform's configuration language is based on a more general
```
Comment (nfagerlund) [inline website/docs/configuration/syntax.html.md]: How's this? (below)

File: website/docs/configuration/types.html.md [hunk 6]
```diff
@@ -0,0 +1,187 @@
+---
+layout: "docs"
+page_title: "Type Constraints - Configuration Language"
+sidebar_current: "docs-config-types"
+description: |-
+  Terraform module authors and provider developers can use detailed type
+  constraints to validate the inputs of their modules and resources.
+---
+
+# Type Constraints
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: (Could use special attention to this whole page, but a few specific things called out below.)

File: website/docs/configuration/types.html.md [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+
+* Type keywords are unquoted symbols that represent a static type.
+* Type constructor functions are unquoted symbols followed by a pair of
+  parentheses, which contain an argument that specifies more information about
+  the type. Without its argument, a type constructor function does not fully
+  represent a type; instead, it represents a _kind_ of similar types.
+
+Type constraints look like other kinds of Terraform
+[expressions](./expressions.html), but are a special syntax. Within the
+Terraform language, they are only valid in the `type` argument of an [input
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: right?

File: website/docs/configuration/types.html.md [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+* `false` converts to `"false"`, and vice-versa
+* `15` converts to `"15"`, and vice-versa
+
+## The "Any" Type
+
+The type keyword `any` is a special type constraint that accepts any value.
+
+## Complex Types
+
+A _complex_ type is a type that groups multiple values into a single value.
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: I grouped these together into a super-category, since I saw the term "complex type" being used elsewhere in this section.

File: website/docs/configuration/types.html.md [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+  constraints to validate the inputs of their modules and resources.
+---
+
+# Type Constraints
+
+Terraform module authors and provider developers can use detailed type
+constraints to validate user-provided values for their input variables and
+resource arguments. This requires some additional knowledge about Terraform's
+type system, but allows you to build a more resilient user interface for your
+modules and resources.
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: (Is this sufficient/accurate indication of whether you need to read this page or can get away with skipping it?)

File: website/docs/configuration/types.html.md [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+different type than `list(number)`, a list of numbers. All elements of a
+collection must always be of the same type.
+
+The three kinds of collection type in the Terraform language are:
+
+* `list(...)`: a sequence of values identified by consecutive whole numbers
+  starting with zero.
+
+    The type keyword `list` is a shorthand for `list(any)`, which accepts any
+    element type as long as every element is the same type.
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: (Did I get these shorthands right?)

File: website/docs/configuration/types.html.md [hunk 11]
```diff
@@ -0,0 +1,187 @@
...
+* `set(...)`: a collection of unique values that do not have any secondary
+  identifiers or ordering.
+
+    The type keyword `set` is a shorthand for `set(any)`, which accepts any
+    element type as long as every element is the same type.
+
+### Structural Types
+
+A _structural_ type allows multiple values of _several distinct types_ to be
+grouped together as a single value. Structural types require a _schema_ as an
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: (introduced the word "schema" here since that seems to be the name for this thing)

File: website/docs/configuration/types.html.md [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+
+The two kinds of structural type in the Terraform language are:
+
+* `object(...)`: a collection of named attributes that each have their own type.
+
+    The schema for object types is `{<KEY> = <TYPE>, <KEY> = <TYPE>, ...}` — a
+    pair of curly braces containing a comma-separated series of `<KEY> = <TYPE>`
+    pairs. Values that match the object type must contain _exactly_ the
+    specified keys (no more and no fewer), and the value for each key must
+    match the specified type for that key.
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: I had to infer this, because it was too late in the day to bug you!!! Did I get this correct, or does it behave differently?

File: website/docs/configuration/types.html.md [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+* `tuple(...)`: a sequence of elements identified by consecutive whole
+  numbers starting with zero, where each element has its own type.
+
+    The schema for tuple types is `[<TYPE>, <TYPE>, ...]` — a pair of square
+    brackets containing a comma-separated series of types. Values that match the
+    tuple type must have _exactly_ the same number of elements (no more and no
+    fewer), and the value in each position must match the specified type for
+    that position.
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: Likewise this one.

File: website/docs/configuration/types.html.md [hunk 14]
```diff
@@ -0,0 +1,187 @@
...
+values, which are described in [Expressions: Literal
+Expressions](./expressions.html#literal-expressions) as "list/tuple" literals
+and "map/object" literals, respectively.
+
+Terraform does _not_ provide any way to directly represent lists, maps, or sets.
+However, due to the automatic conversion of complex types (described below), the
+difference between similar complex types is almost never relevant to a normal
+user, and most of the Terraform documentation conflates lists with tuples and
+maps with objects. The distinctions are only useful when restricting input
+values for a module or resource.
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: I think I got this messaging right...

File: website/docs/configuration/types.html.md [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+
+* Whenever possible, Terraform converts values between similar kinds of complex
+  types if the provided value is not the exact type requested. "Similar kinds"
+  is defined as follows:
+    * Objects and maps are similar.
+    * Tuples and lists are similar.
+    * Sets are _almost_ similar to both tuples and lists. When a value is
+      converted to a set, duplicate values are discarded and the ordering of
+      elements is lost. When a set is converted to another type, no particular
+      ordering of elements is guaranteed.
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: right?

File: website/docs/configuration/types.html.md [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+  in [Conversion of Primitive Types](#conversion-of-primitive-types).
+
+For example: if a module argument requires a value of type `list(string)` and a
+user provides the tuple `["a", 15, true]`, Terraform will internally transform
+the value to `["a", "15", "true"]` by converting the elements to the required
+`string` element type. Later, if the module uses those elements to set different
+resource arguments that require a string, a number, and a bool (respectively),
+Terraform will automatically convert the second and third strings to the
+required types at that time, since they contain valid representations of a
+number and a bool.
```
Comment (nfagerlund) [inline website/docs/configuration/types.html.md]: I think this is aligned with the example you walked me through, but I might have forgotten some subtlety.

File: website/layouts/functions.erb [hunk 17]
```diff
@@ -5,14 +5,14 @@
...
           <li<%= sidebar_current("docs-config-functions") %>>
             <a class="back" href="/docs/configuration/functions.html">Functions</a>
           </li>
-  
+
           <li<%= sidebar_current("docs-funcs-numeric") %>>
-            <a href="#">Numeric Functions</a>
-            <ul class="nav nav-visible">
+            <a href="#docs-funcs-numeric">Numeric Functions</a>
+            <ul class="nav nav-visible" id="docs-funcs-numeric">
```
Comment (nfagerlund) [inline website/layouts/functions.erb]: (this is just to fix an annoying behavior where it would jump to the top of the page and lose your spot if you clicked on one of these "header" links in the nav.)

File: website/docs/configuration/expressions.html.md [hunk 18]
```diff
@@ -16,195 +16,204 @@ references to data exported by resources, arithmetic, conditional evaluation,
...
-  numbers like `15` and fractional values such as `6.283185`.
-
+  numbers like `15` and fractional values like `6.283185`.
 * `bool`: either `true` or `false`. `bool` values can be used in conditional
   logic.
+* `list` (or `tuple`): a sequence of values, like
+  `["us-west-1a", "us-west-1c"]`. Elements in a list or tuple are identified by
+  consecutive whole numbers, starting with zero.
+* `map` (or `object`): a group of values identified by named labels, like
+  `{name = "Mabel", age = 52}`.
```
Comment (nfagerlund) [inline website/docs/configuration/expressions.html.md]: This is the spot where I started conflating the similar complex types more aggressively, pushing the full details out into the type constraints page.

File: website/docs/configuration/expressions.html.md [hunk 19]
```diff
@@ -16,195 +16,204 @@ references to data exported by resources, arithmetic, conditional evaluation,
...
+"map").
 
-There is no direct syntax for creating collection type values, but the
-Terraform language can automatically convert a structural type value (as
-defined in the next section) to a similar collection type as long as all
-of its elements can be converted to the required element type.
+However, module authors and provider developers should understand the
+differences between these similar types (and the related `set` type), since they
+offer different ways to restrict the allowed values for input variables and
+resource arguments.
```
Comment (nfagerlund) [inline website/docs/configuration/expressions.html.md]: And this is the spot where I nudge you to go learn the differences, if you're one of the people who actually needs to know. (There's a link on the next line, but the diff interleaved a big block of removed text here, sorry.)

File: website/docs/configuration/expressions.html.md [hunk 20]
```diff
@@ -16,195 +16,204 @@ references to data exported by resources, arithmetic, conditional evaluation,
...
+    * See [String Literals](#string-literals) below for information about escape
+      sequences and the heredoc syntax.
+    * See [String Templates](#string-templates) below for information about
+      interpolation and template directives.
+* Numbers are represented by unquoted sequences of digits with or without a
+  decimal point, like `15` or `6.283185`.
+* Bools are represented by the unquoted symbols `true` and `false`.
+* The null value is represented by the unquoted symbol `null`.
+* Lists/tuples are represented by a pair of square brackets containing a
+  comma-separated sequence of values, like `["a", 15, true]`.
```
Comment (nfagerlund) [inline website/docs/configuration/expressions.html.md]: Another important spot where I conflated list/tuples etc. a bit more for convenience.

File: website/docs/configuration/types.html.md (hunk 15 above)
Comment (apparentlymart) [reply website/docs/configuration/types.html.md]: This is accurate.

We _do_ guarantee ordering of `set(string)` in particular to be lexicographical ordering by the string values, but perhaps that special case is better noted somewhere else. (Not sure where, though.)

File: website/docs/configuration/types.html.md (hunk 16 above)
Comment (apparentlymart) [reply website/docs/configuration/types.html.md]: This is all correct.

File: website/docs/configuration/variables.html.md [hunk 21]
```diff
@@ -10,17 +10,24 @@ description: |-
...
 
 Input variable usage is introduced in the Getting Started guide section
 [_Input Variables_](/intro/getting-started/variables.html).
 
+-> **Note:** For brevity, input variables are often referred to as just
+"variables" or "Terraform variables" when it is clear from context what sort of
+variable is being discussed. Other kinds of variables in Terraform include
+_environment variables_ (set by the shell where Terraform runs) and _expression
+variables_ (used to indirectly represent a value in an
+[expression](./expressions.html)).
```
Comment (apparentlymart) [inline website/docs/configuration/variables.html.md]: :+1: this additional context is very helpful

File: website/docs/configuration/variables.html.md [hunk 22]
```diff
@@ -134,35 +148,42 @@ commentary for module maintainers, use comments.
...
-that is accepted by the `terraform plan` and `terraform apply` commands:
+To specify individual modules on the command line, use the `-var` option
+when running the `terraform plan` and `terraform apply` commands:
 
 ```
 terraform apply -var="image_id=ami-abc123"
 ```
 
-### Variable Definitions Files
+The `-var` argument can be used any number of times in a single command.
```
Comment (apparentlymart) [inline website/docs/configuration/variables.html.md]: Since you changed "argument" to "option" above (which makes sense to me!) we should probably change this one too.

File: website/docs/configuration/resources.html.md [hunk 23]
```diff
@@ -46,21 +46,37 @@ arguments defined specifically for [the `aws_instance` resource type](/docs/prov
...
+and apply across all resource types. (See [Meta-Arguments](#meta-arguments) below.)
+
+### Documentation for Resource Types
+
+[Terraform's provider documentation][providers] is the primary place to
+learn which resource types are available and which arguments to use for each
+resource type. Once you understand Terraform's basic syntax, the provider
+documentation will be where you spend the majority of your time on this website.
+
+The "[Providers][]" link at the top level of the navigation sidebar will take
```
Comment (apparentlymart) [inline website/docs/configuration/resources.html.md]: I'm not familiar with this link syntax... does this mean to use the link text itself (case-insensitively, I assume) to find the named link target further down the page? Good if so, just want to make sure!

File: website/docs/configuration/data-sources.html.md [hunk 24]
```diff
@@ -125,7 +125,7 @@ resources.
 
 ## Multiple Resource Instances
 
-Data resources support [the `count` meta-argument](/docs/configuration/resources.html#multiple-resource-instances)
+Data resources support [the `count` meta-argument](./resources.html#multiple-resource-instances)
```
Comment (apparentlymart) [inline website/docs/configuration/data-sources.html.md]: I think this fragment will now be `#count-multiple-resource-instances` after the changes to headings on the resources page. Perhaps this applies to the other meta-argument links on this page too.

File: website/docs/configuration/types.html.md [hunk 25]
```diff
@@ -0,0 +1,187 @@
...
+    The type keyword `list` is a shorthand for `list(any)`, which accepts any
+    element type as long as every element is the same type.
+* `map(...)`: a collection of values where each is identified by a string label.
+
+    The type keyword `map` is a shorthand for `map(any)`, which accepts any
+    element type as long as every element is the same type.
+* `set(...)`: a collection of unique values that do not have any secondary
+  identifiers or ordering.
+
+    The type keyword `set` is a shorthand for `set(any)`, which accepts any
```
[the remark this answers is not in the document]
Comment (nfagerlund) [reply website/docs/configuration/types.html.md]: agree: with the caveat added, it isn't confusing for `set` to be missing this.

File: website/docs/configuration/types.html.md (hunk 15 above)
Comment (nfagerlund) [reply website/docs/configuration/types.html.md]: This is probably the place for it. 👍🏼

File: website/docs/configuration/data-sources.html.md (hunk 24 above)
Comment (nfagerlund) [reply website/docs/configuration/data-sources.html.md]: Blaugh, I think I fixed it once and then unfixed it during an interactive rebase. Let's try this again!

Comment (nfagerlund) [reply website/docs/configuration/data-sources.html.md]: Oh, nope, I actually only fixed half of them originally 😅 Thanks for the catch!

File: website/docs/configuration/functions/substr.html.md [hunk 26]
```diff
@@ -1,6 +1,6 @@
 ---
 layout: "functions"
-page_title: "substr function"
+page_title: "substr - Functions - Configuration Language"
```
Comment (nfagerlund) [inline website/docs/configuration/functions/substr.html.md]: In retrospect, I really should have done these function page titles in a separate PR to cut down on the noise here. 😑

File: website/docs/configuration/index.html.md [hunk 27]
```diff
@@ -16,23 +16,52 @@ describing an intended goal rather than the steps to reach that goal.
...
+  <IDENTIFIER> = <EXPRESSION> # Argument
+}
+```
+
+- _Blocks_ are containers for other content and usually represent the
+  configuration of some kind of object, like a resource. Blocks have a
+  _block type,_ can have zero or more _labels,_ and have a _body_ that contains
+  any number of arguments and nested blocks. Most of Terraform's features are
+  controlled by top-level blocks in a configuration file.
+- _Arguments_ assign a value to a name. They almost always occur within blocks.
```
[the remark this answers is not in the document]
Comment (nfagerlund) [reply website/docs/configuration/index.html.md]: Oh thanks! That's exactly what I was wondering, was whether we were used to thinking of `.tfvars` as being part of the Terraform language. Will adjust.

File: website/docs/configuration/resources.html.md (hunk 23 above)
Comment (nfagerlund) [reply website/docs/configuration/resources.html.md]: @apparentlymart Yeah! The original markdown spec calls this the ["the implicit link name shortcut"](https://daringfireball.net/projects/markdown/syntax#link) (cmd-f for the name, that page doesn't have very good linkable anchors), and the GFM spec calls it a ["collapsed reference link"](https://github.github.com/gfm/#collapsed-reference-link). In every implementation I've seen, they act case-insensitively, which is convenient.

File: website/docs/configuration/syntax.html.md [hunk 28]
```diff
@@ -60,106 +76,45 @@ resource "aws_instance" "example" {
...
-In practice, this means that identifiers can contain letters, digits,
-underscores, and hyphens. To avoid ambiguity with literal numbers, the
-first character of an identifier must not be a digit.
+### Short Rules for Identifiers
+
+Identifiers can contain letters, digits, underscores (`_`), and hyphens (`-`).
+The first character of an identifier must not be a digit, to avoid ambiguity
+with literal numbers.
+
+### Complete Rules for Identifiers
```
[the remark this answers is not in the document]
Comment (nfagerlund) [reply website/docs/configuration/syntax.html.md]: Yeah, you're probably right; I veered a smidge too far in signposting that it was ok to ignore the second paragraph.
